### PR TITLE
[bitnami/minio] Add support to deploy MinIO as a Gateway

### PR DESCRIPTION
**Pull request by juan131**
Signed-off-by: juan131 <juanariza@vmware.com>

**Description of the change**

This PR adds support to deploy MinIO as a Gateway for other storage systems such as AWS S3, NAS, Azure Blob Storage or Google Cloud Storage.

**Benefits**

- Flexibility
- New feature

**Possible drawbacks**

None

**Applicable issues**

  - fixes https://github.com/bitnami/charts/issues/1698

**Additional information**

Some references:

- https://docs.minio.io/docs/minio-gateway-for-azure
- https://docs.minio.io/docs/minio-gateway-for-gcs
- https://docs.minio.io/docs/minio-gateway-for-nas
- https://docs.minio.io/docs/minio-gateway-for-s3

**Checklist** 

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)

### CODE REVIEW

File: bitnami/minio/templates/gateway/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+          {{- if .Values.resources }}
+          resources: {{- toYaml .Values.resources | nindent 12 }}
+          {{- end }}
+          volumeMounts:
+            {{- if and (eq .Values.gateway.type "gcs") .Values.gateway.auth.gcs.keyJson }}
+            - name: minio-credentials
+              mountPath: /opt/bitnami/minio/secrets/
+              readOnly: true
+            {{- end }}
+            {{- if and .Values.gateway.enabled (eq .Values.gateway.type "nas") }}
```
Comment (javsalgar) [inline bitnami/minio/templates/gateway/deployment.yaml:136]: No need to check if gateway.enabled, it wouldn't enter the initial if

File: bitnami/minio/README.md [hunk 2]
```diff
@@ -208,7 +208,7 @@ The following table lists the configurable parameters of the MinIO&reg; chart an
 
 | Parameter                                 | Description                                                                         | Default                                                      |
```
Comment (javsalgar) [inline bitnami/minio/README.md:209]: Not related to you, but could you add these missing parameters?

```
statefulset.podManagementPolicy
containerPort
ingress.extraPaths
persistence.accessModes
pdb.create
pdb.minAvailable
```

File: bitnami/minio/templates/gateway/pvc.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{{- if and .Values.gateway.enabled (eq .Values.gateway.type "nas") (include "minio.createPVC" .) }}
```
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/minio/templates/gateway/pvc.yaml:1]: Good point

File: bitnami/minio/values.yaml [hunk 4]
```diff
@@ -164,10 +164,10 @@ args: []
 # schedulerName: stork
 
 ## MinIO(R) deployment parameters
-## Only when mode is 'standalone'
+## Only when 'mode' is 'standalone' or 'gateway.enabled' is 'true'
 ##
 deployment:
```
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/minio/values.yaml:169]: Correct, I don't like it; but I don't want to include a breaking change.

File: bitnami/minio/templates/standalone/deployment.yaml [hunk 5]
```diff
@@ -104,23 +104,23 @@ spec:
               value: {{ ternary "yes" "no" .Values.forceNewKeys | quote }}
             {{- if .Values.useCredentialsFile }}
             - name: MINIO_ACCESS_KEY_FILE
-              value: "/opt/bitnami/minio/secrets/access-key"
+              value: "/opt/bitnami/minio/secrets/{{ include "minio.secret.userKey" . }}"
```
[the remark this answers is not in the document]
Comment (juan131) [reply bitnami/minio/templates/standalone/deployment.yaml:107]: I'll check/adapt it

**Review by javsalgar — APPROVED** [review]

**Comment by maxisam** [comment]
Is it possible to use custom acess/secret keys?

https://docs.min.io/docs/minio-gateway-for-azure

It seems like it doesn't work in current setup

**Comment by maxisam** [comment]
Here are things I found we can improve for gateway.

1. If you use existingSecret, it ignore the key from gateway section. 
For Azure the keys for secret are azure-storage-account-name & azure-storage-account-key

2. Currently we can use extraEnvVarsSecret to set env variable for azure. 

Anyway, thanks for putting this together.

**Comment by juan131** [comment]
Thanks for the feedback @maxisam 

I agree that we should provide a way to use custom access/secret keys. Let me work on a PR to fix that.
I will also include the missing `MINIO_PROMETHEUS_AUTH_TYPE` environment variable.

**Comment by juan131** [comment]
See https://github.com/bitnami/charts/pull/5833

**Comment by maxisam** [comment]
@juan131  while you are at it, you might wanna make prometheus path as a variable. They have deprecated the one in the chart. Thanks for the fast response!

**Comment by juan131** [comment]
What do you exactly mean @maxisam ? I don't see any environment variable documented at MinIO docs to modify the path, see:

- https://docs.min.io/docs/how-to-monitor-minio-using-prometheus.html

**Comment by maxisam** [comment]
Sorry, I meant we should provide a way to let people use the new path.

https://github.com/bitnami/charts/blob/d0da08a308e3c1802ed7b102d3a93589bbfda6b6/bitnami/minio/templates/servicemonitor.yaml#L24

**Comment by juan131** [comment]
Do you mean `/minio/v2/metrics/cluster` @maxisam ?

**Comment by maxisam** [comment]
Yes

**Comment by juan131** [comment]
@maxisam see https://github.com/bitnami/charts/pull/5875

**Pull request merged**